File: WebSocketUnityClient/Assets/WebSocketClient/Runtime/Scripts/Logic/ConnMonitor.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;
using System.Threading;
using UIFramework;

namespace WebSocketClient
{
    /// <summary>
    /// websocket 网络连接业务逻辑管理
    /// 1.如果网络心跳超时，复制通知 server 关闭连接
    /// </summary>
    public class ConnMonitor
    {
        private const string serviceName = "ConnMonitorService";

        private WSBackend _backend;

        private const int _intervalMS = 5000;

        public void Init()
        {
            _backend.OnBackendStateChanged += Singleton_OnBackendStateChanged;
            _backend.OnBackendResponse += Backend_OnBackendResponse;
            _backend.OnBackendNotify += Backend_OnBackendNotify;
        }

        public void Release()
        {
            _backend.OnBackendStateChanged -= Singleton_OnBackendStateChanged;
            _backend.OnBackendResponse -= Backend_OnBackendResponse;
            _backend.OnBackendNotify -= Backend_OnBackendNotify;
        }

        private void Backend_OnBackendResponse(ResponsePack resp)
        {
            StartNewHeartBeatTick();
        }

        private void Backend_OnBackendNotify(NotifyPack not)
        {
            StartNewHeartBeatTick();
        }

        /// <summary>
        /// 收到任何消息后，需要重置心跳时间，并且终端当前的心跳计时
        /// </summary>
        private CancellationTokenSource _heartBeatCTS;

        private void StartNewHeartBeatTick()
        {
            if (_heartBeatCTS != null)
            {
                _heartBeatCTS.Cancel();
                _heartBeatCTS.Dispose();
            }
            _heartBeatCTS = new CancellationTokenSource();
            Task.Run(async () =>
            {
                while (_heartBeatCTS != null && !_heartBeatCTS.IsCancellationRequested)
                {
                    // Delay before sending ping
                    await Task.Delay(_intervalMS, _heartBeatCTS.Token);

                    // Check cancellation before sending ping
                    if (_heartBeatCTS.IsCancellationRequested)
                    {
                        break;
                    }

                    // Send ping
                    Utility.LogDebug("ConnMonitor", "send wsping");
                    await _backend.CreateBackendRequest(serviceName, BackendOps.WSPing, null);
                }
            });
        }

        private void Singleton_OnBackendStateChanged()
        {
            if (_backend.State == WSBackend.WSBackendState.Open)
            {
                Utility.LogDebug("ConnMonitor", "heart beat monitor start");
                StartNewHeartBeatTick();
            }
            else
            {
                _heartBeatCTS.Cancel();
                _heartBeatCTS.Dispose();
                _heartBeatCTS = null;
                Utility.LogDebug("ConnMonitor", "heart beat monitor cancel by ws close");
            }
        }

        public static ConnMonitor Create(WSBackend backend)
        {
            ConnMonitor instance = new ConnMonitor();
            instance._backend = backend;
            return instance;
        }
    }
}

File: WebsocketServer/WebsocketServer/wwwroot/TypeScriptClient/lib/clientKernal/ConnMonitor.js
var WebsocketTSClient;
(function (WebsocketTSClient) {
    /// send heart beat to server
    class ConnMonitor {
        constructor() {
            this.Singleton_OnBackendStateChanged = () => {
                if (this._backend.IsConnected) {
                    clearTimeout(this._heartBeatCTS);
                    WebsocketTSClient.Utility.LogDebug("[ConnMonitor]", "heart beat monitor start");
                    this.StartNewHeartBeatTick();
                }
                else {
                    clearTimeout(this._heartBeatCTS);
                    this._heartBeatCTS = null;
                    WebsocketTSClient.Utility.LogDebug("[ConnMonitor]", "heart beat monitor cancel by ws close");
                }
            };
        }
        Init() {
            this._backend.OnStateChanged.AddListener(this.Singleton_OnBackendStateChanged);
            //this._backend.OnResponse.AddListener(this.Backend_OnBackendResponse);
            //this._backend.OnNotify.AddListener(this.Backend_OnBackendNotify);
        }
        Release() {
            this._backend.OnStateChanged.RmListener(this.Singleton_OnBackendStateChanged);
            //this._backend.OnResponse.RmListener(this.Backend_OnBackendResponse);
            //this._backend.OnNotify.RmListener(this.Backend_OnBackendNotify);
        }
        //private Backend_OnBackendResponse = (resp: ResponsePack): void =>
        //{
        //    this.StartNewHeartBeatTick();
        //}
        //private Backend_OnBackendNotify = (not: NotifyPack): void =>
        //{
        //    this.StartNewHeartBeatTick();
        //}
        StartNewHeartBeatTick() {
            if (!this._backend.IsConnected) {
                return;
            }
            WebsocketTSClient.Utility.LogDebug("[ConnMonitor]", "send WSPing");
            WebsocketTSClient.BackendRequest.CreateRetry(ConnMonitor.serviceName, "WSPing", null, null, (errCode, data, context) => {
                if (errCode == WebsocketTSClient.ErrCode.OK) {
                    this._heartBeatCTS = setTimeout(() => {
                        this.StartNewHeartBeatTick();
                    }, ConnMonitor.intervalMS);
                }
                else {
                    // print error message
                    WebsocketTSClient.Utility.LogDebug("[ConnMonitor]", "send WSPing error " + errCode);
                }
            });
        }
        static Create(backend) {
            const instance = new ConnMonitor();
            instance._backend = backend;
            return instance;
        }
    }
    ConnMonitor.serviceName = "ConnMonitorService";
    ConnMonitor.intervalMS = 5000;
    WebsocketTSClient.ConnMonitor = ConnMonitor;
})(WebsocketTSClient || (WebsocketTSClient = {}));
